contrib/slurm/slurmexecutor.py

```diff
@@ -0,0 +1,297 @@
+            self.run_finished_callback()
+            _Worker.working_queue.task_done()
+
+    def execute(self, run):
+        """
+        This function executes the tool with a sourcefile with options.
+        It also calls functions for output before and after the run.
+        """
+        self.output_handler.output_before_run(run)
+        benchmark = self.benchmark
+
+        args = run.cmdline()
+        logging.debug("Command line of run is %s", args)
+
+        try:
+            with open(run.log_file, "w") as f:
+                for i in range(6):
+                    f.write(os.linesep)
+
+            timelimit = self.benchmark.rlimits.cputime
+
+            run_result = run_slurm(
+                benchmark,
+                args,
+                run.log_file,
+                timelimit,
+                self.my_cpus,
+                benchmark.rlimits.memory,
+            )
+
+        except KeyboardInterrupt:
+            # If the run was interrupted, we ignore the result and cleanup.
+            stop()
+
+        if STOPPED_BY_INTERRUPT:
+            try:
+                if benchmark.config.debug:
+                    os.rename(run.log_file, run.log_file + ".killed")
+                else:
+                    os.remove(run.log_file)
+            except OSError:
+                pass
+            return 1
+
+        run.set_result(run_result)
+        self.output_handler.output_after_run(run)
+        return None
+
+
+def run_slurm(benchmark, args, log_file, timelimit, cpus, memory):
+    srun_timelimit_h = int(timelimit / 3600)
+    srun_timelimit_m = int((timelimit % 3600) / 60)
+    srun_timelimit_s = int(timelimit % 60)
+    srun_timelimit = f"{srun_timelimit_h}:{srun_timelimit_m}:{srun_timelimit_s}"
+
+    mem_per_cpu = int(memory / cpus / 1000000)
+
+    assert benchmark.config.scratchdir and os.path.exists(benchmark.config.scratchdir)
+
+    with tempfile.TemporaryDirectory(dir=benchmark.config.scratchdir) as tempdir:
+
+        os.makedirs(os.path.join(tempdir, "upper"))
+        os.makedirs(os.path.join(tempdir, "work"))
+
+        tool_command = " ".join(args)
+        singularity_command = (
+            f"singularity exec "
+            f"-B $PWD:/lower --no-home "
+            f"-B {tempdir}:/overlay "
+            f'--fusemount "container:fuse-overlayfs -o lowerdir=/lower -o upperdir=/overlay/upper -o workdir=/overlay/work $HOME" '
+            f"{benchmark.config.singularity} {tool_command}"
+            if benchmark.config.singularity
+            else tool_command
+        )
+        srun_command = (
+            f"srun "
+            f"-t {srun_timelimit} "
+            f"-c {cpus} "
+            f"-o {log_file} "
+            f"--mem-per-cpu {mem_per_cpu} "
+            f"--threads-per-core=1 "
+            f"--ntasks=1 "
+            f"{singularity_command}"
+        )
+        jobid_command = (
+            f"{srun_command} 2>&1 | grep -o 'job [0-9]* queued' | grep -o '[0-9]*'"
```

Comment:
I would prefer if we could get rid of the pipes and the shell around it, because that would make the quoting easier and also be more robust in general. For example, right now you loose the exit code from `seff`.

Reply:
Sure, I'll try to rework this then.

Reply:
Without the pipe and with the added quoting this is already better. But are still shell invocations and for example more special characters (like quotes) will still break stuff.

My goal would be to get rid of the shell completely. Only if we really need it, it may stay, but then we need to be really really careful with quoting or escaping.

Also, when chaining commands with options like in `foo --abc bar --def` one needs to take care that `foo` handles its own options and the ones it should pass through correctly. Do `srun`, `singularity` and `seff` do this automatically, even if for example the final tool gets a command-line argument that happens to have a name collision with one of the wrapper command's parameters? Or is there something that we can do to improve that?
For example, it is a common convention that such wrapper tools accept `--` as a delimiter between own arguments and the rest. Even `runexec` supports that, you can do `runexec ... (runexec parameters) -- tool ... (tool parameters)`. If the involved tools here also support this, it may be good to add this.

Reply:
i'll try to research this, I frankly don't know how they handle such params.